### PR TITLE
`_launch` refactor and types [1/n]

### LINKED ISSUE #7242: [Trainer] Use name mangling to allow for subclassing
## 🐛 Bug

The `Trainer` class uses its `self.fit` directly in several places. That causes problems when the `Trainer` class is subclassed and the behavior of the `fit` method altered. As it's now impossible to call the `test` method because it uses `self.fit` internally. 

Since things like `trainer.validate` and `GridSearch` are not yet implemented in `PyTorchLightning` the user should be allowed to subclass and modify those methods without worrying about breaking the `Trainer` code.

## Please reproduce using the BoringModel

https://colab.research.google.com/drive/1jcq5g0XF_3xTLR0p7ZuEMfHcowtZhf4j?usp=sharing

### Expected behavior

`SeedSearchTrainer` uses the `fit` method of the base class (`pl.Trainer`) when `test` method of the base class is called.

### Additional context

This could be solved using name mangling. 

Store a local reference to the `fit` method by adding a `__fit` attribute to the `Trainer` class.
```python
class Trainer():
    ...
    __fit = fit
``` 

Then on lines:
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/trainer.py#L920
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/trainer.py#L981
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/trainer.py#L1075

replace `self.fit` by `self.__fit`

What this achieves is that whenever the `self.__fit` method is called the name mangling would change its name to `_Trainer__fit` to assure that the method of the `Trainer` class is called rather than some overwritten version of the method from the derived class.

This approach can also be implemented for other methods that the user may want to subclass and that are used internally in the base class as `self.method`.

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

- Split `fit()` and `_launch()` to avoid `if` checks for state in `fit()`. Suggested by @ananthsub: https://github.com/PyTorchLightning/pytorch-lightning/pull/7173#discussion_r618895633
- Typing
- Fixed bug where `stage` was not being set back to `tuning` after the tuner calls `fit` internally

Fixes #7242 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [n/a] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 1]
```diff
@@ -123,30 +124,21 @@ def lightning_module(self) -> 'pl.LightningModule':
         """Returns the pure LightningModule without potential wrappers"""
         return unwrap_lightning_module(self._model)
 
-    @property
-    def results(self) -> Any:
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:127]: please leave it. read only is better and I can improve the docs. this is only here for handling the ddp-spawn case internally.

Comment (carmocca) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:127]: Done

File: pytorch_lightning/tuner/batch_size_scaling.py [hunk 2]
```diff
@@ -218,7 +218,7 @@ def _run_binsearch_scaling(trainer, model, new_size, batch_arg_name, max_trials,
         trainer.global_step = 0  # reset after each try
         try:
             # Try fit
-            trainer.fit(model, **fit_kwargs)
+            trainer.tuner._fit(model, **fit_kwargs)
```
Comment (ananthsub) [inline pytorch_lightning/tuner/batch_size_scaling.py:221]: it's odd we call a private method like this

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/tuner/batch_size_scaling.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/tuner/batch_size_scaling.py:221]: I know... I added the `_` so users don't rely on it, not because I don't want another of our components calling it.

I guess this is the difference between:
- `_foo`: discourage users
- `__foo`: actual private method

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -860,14 +830,50 @@ def run_sanity_check(self, ref_model):
             # prevents sanity check to affect random sampling in training
             reset_seed()
 
+    def fit(
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:833]: don't we want to move this up? It would be nice if the entry points fit, validate, test came first.

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:833]: Agree. Let's try to organise, so it is simple to follow.

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -408,36 +409,13 @@ def __init__(
         # Callback system
         self.on_init_end()
 
-    def fit(
+    def _fit_impl(
         self,
         model: LightningModule,
         train_dataloader: Any = None,
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:415]: Should we remove `train_dataloader ` and `val_dataloaders` and `damodule` and `self.train_loop.setup_fit(model, train_dataloader, val_dataloaders, datamodule)` inside of fit function.

Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:415]: Would `dispatch_laucher` be a better name than `_fit_impl` ? Or something else. This function does more than fitting.

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:833]: I'll do this in a separate PR. Keeping them where they are for now so they are together

File: pytorch_lightning/trainer/trainer.py (hunk 4 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py]: Done both comments. Called it `_launch` instead

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -860,14 +827,63 @@ def run_sanity_check(self, ref_model):
...
+
+            datamodule: A instance of :class:`LightningDataModule`.
+        """
+        Trainer._log_api_event("fit")
+
+        self.state = TrainerState.FITTING
+        self.training = True
+
+        # if a datamodule comes in as the second arg, then fix it for the user
+        if isinstance(train_dataloader, LightningDataModule):
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:858]: Neat !

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -408,36 +409,13 @@ def __init__(
         # Callback system
         self.on_init_end()
 
-    def fit(
+    def _launch(
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:412]: Nit : I prefer _run or _execute vs launch, which I associate more with distributed launch/process creation

Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:412]: Bit late as it is already merged, but if you want to open a PR with the renaming change, that's fine by me! 👍
